File: slicing-ios/SlicingGame/Components/Game/LevelSlicePreviewView.swift
//
//  LevelSlicePreviewView.swift
//  Game view: shows a helper line for previewing a slice
//

import UIKit
import UniLayout
import JsonInflator

class LevelSlicePreviewView: UniView {
    
    // --
    // MARK: Members
    // --
    
    private var sliceVector: Vector?
    

    // --
    // MARK: Viewlet integration
    // --

    class func viewlet() -> JsonInflatable {
        return ViewletClass()
    }
    
    private class ViewletClass: JsonInflatable {
        
        func create() -> Any {
            return LevelSlicePreviewView()
        }
        
        func update(convUtil: InflatorConvUtil, object: Any, attributes: [String: Any], parent: Any?, binder: InflatorBinder?) -> Bool {
            if let slicePreview = object as? LevelSlicePreviewView {
                // Apply start position
                let startPositionArray = convUtil.asDimensionArray(value: attributes["start"])
                if startPositionArray.count == 2 {
                    slicePreview.start = CGPoint(x: startPositionArray[0], y: startPositionArray[1])
                } else {
                    slicePreview.start = nil
                }

                // Apply end position
                let endPositionArray = convUtil.asDimensionArray(value: attributes["end"])
                if endPositionArray.count == 2 {
                    slicePreview.end = CGPoint(x: endPositionArray[0], y: endPositionArray[1])
                } else {
                    slicePreview.end = nil
                }
                
                // Apply colors
                slicePreview.color = convUtil.asColor(value: attributes["color"])
                slicePreview.stretchedColor = convUtil.asColor(value: attributes["stretchedColor"])

                // Generic view properties
                ViewletUtil.applyGenericViewAttributes(convUtil: convUtil, view: slicePreview, attributes: attributes)
                return true
            }
            return false
        }
        
        func canRecycle(convUtil: InflatorConvUtil, object: Any, attributes: [String: Any]) -> Bool {
            return type(of: object) == LevelSlicePreviewView.self
        }
        
    }
    
    
    // --
    // MARK: Initialization
    // --
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        setup()
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        setup()
    }
    
    private func setup() {
        // No implementation
    }
    
    
    // --
    // MARK: Configurable values
    // --
    
    var start: CGPoint? {
        didSet {
            if start != oldValue {
                if let start = start, let end = end {
                    sliceVector = Vector(start: start, end: end)
                } else {
                    sliceVector = nil
                }
                updateLayers()
            }
        }
    }

    var end: CGPoint? {
        didSet {
            if end != oldValue {
                if let start = start, let end = end {
                    sliceVector = Vector(start: start, end: end)
                } else {
                    sliceVector = nil
                }
                updateLayers()
            }
        }
    }
    
    var color: UIColor? {
        didSet {
            updateLayers()
        }
    }

    var stretchedColor: UIColor? {
        didSet {
            updateLayers()
        }
    }

    
    // --
    // MARK: Update layers
    // --
    
    private func updateLayers() {
        // Clear sublayers
        for layer in layer.sublayers ?? [] {
            layer.removeFromSuperlayer()
        }

        // Add stretched line to component edges
        if let sliceVector = sliceVector, sliceVector.isValid() {
            let stretchedVector = sliceVector.stretchedToEdges(topLeft: CGPoint(x: 0, y: 0), bottomRight: CGPoint(x: bounds.width, y: bounds.height))
            if stretchedVector.isValid() {
                let stretchedLine = CAShapeLayer()
                let stretchedLinePath = UIBezierPath()
                stretchedLinePath.move(to: stretchedVector.start)
                stretchedLinePath.addLine(to: stretchedVector.end)
                stretchedLine.path = stretchedLinePath.cgPath
                stretchedLine.strokeColor = stretchedColor?.cgColor
                stretchedLine.lineWidth = AppDimensions.slicePreviewStretchedWidth
                layer.addSublayer(stretchedLine)
            }
        }
        
        // Add start/end dots
        let dotOffset = AppDimensions.slicePreviewDot / 2
        if let start = start {
            let startDot = CAShapeLayer()
            startDot.path = UIBezierPath(ovalIn: CGRect(x: start.x - dotOffset, y: start.y - dotOffset, width: AppDimensions.slicePreviewDot, height: AppDimensions.slicePreviewDot)).cgPath
            startDot.fillColor = color?.cgColor
            layer.addSublayer(startDot)
        }
        if let end = end {
            let endDot = CAShapeLayer()
            endDot.path = UIBezierPath(ovalIn: CGRect(x: end.x - dotOffset, y: end.y - dotOffset, width: AppDimensions.slicePreviewDot, height: AppDimensions.slicePreviewDot)).cgPath
            endDot.fillColor = color?.cgColor
            layer.addSublayer(endDot)
        }
        
        // Add line
        if let sliceVector = sliceVector, sliceVector.isValid() {
            let line = CAShapeLayer()
            let linePath = UIBezierPath()
            let dash = NSNumber(value: Float(AppDimensions.slicePreviewDash))
            linePath.move(to: sliceVector.start)
            linePath.addLine(to: sliceVector.end)
            line.path = linePath.cgPath
            line.strokeColor = color?.cgColor
            line.lineWidth = AppDimensions.slicePreviewWidth
            line.lineDashPattern = [dash, dash]
            layer.addSublayer(line)
        }
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        updateLayers()
    }

}

File: slicing-android/app/src/main/java/com/crescentflare/slicinggame/components/game/LevelSlicePreviewView.kt
package com.crescentflare.slicinggame.components.game

import android.annotation.TargetApi
import android.content.Context
import android.graphics.*
import android.os.Build
import android.util.AttributeSet
import com.crescentflare.jsoninflator.JsonInflatable
import com.crescentflare.jsoninflator.binder.InflatorBinder
import com.crescentflare.jsoninflator.utility.InflatorMapUtil
import com.crescentflare.slicinggame.R
import com.crescentflare.slicinggame.components.utility.ViewletUtil
import com.crescentflare.slicinggame.infrastructure.geometry.Vector
import com.crescentflare.unilayout.containers.UniFrameContainer


/**
 * Game view: shows a helper line for previewing a slice
 */
open class LevelSlicePreviewView : UniFrameContainer {

    // --
    // Static: viewlet integration
    // --

    companion object {

        val viewlet: JsonInflatable = object : JsonInflatable {
            override fun create(context: Context): Any {
                return LevelSlicePreviewView(context)
            }

            override fun update(mapUtil: InflatorMapUtil, obj: Any, attributes: Map<String, Any>, parent: Any?, binder: InflatorBinder?): Boolean {
                if (obj is LevelSlicePreviewView) {
                    // Apply start position
                    val startPositionList = mapUtil.optionalDimensionList(attributes, "start")
                    if (startPositionList.size == 2) {
                        obj.start = Point(startPositionList[0], startPositionList[1])
                    } else {
                        obj.start = null
                    }

                    // Apply end position
                    val endPositionList = mapUtil.optionalDimensionList(attributes, "end")
                    if (endPositionList.size == 2) {
                        obj.end = Point(endPositionList[0], endPositionList[1])
                    } else {
                        obj.end = null
                    }

                    // Apply colors
                    obj.color = mapUtil.optionalColor(attributes, "color", Color.TRANSPARENT)
                    obj.stretchedColor = mapUtil.optionalColor(attributes, "stretchedColor", Color.TRANSPARENT)

                    // Generic view properties
                    ViewletUtil.applyGenericViewAttributes(mapUtil, obj, attributes)
                    return true
                }
                return false
            }

            override fun canRecycle(mapUtil: InflatorMapUtil, obj: Any, attributes: Map<String, Any>): Boolean {
                return obj::class == LevelSlicePreviewView::class
            }
        }
    }


    // --
    // Members
    // --

    private val paint = Paint(Paint.ANTI_ALIAS_FLAG)
    private val dotSize = resources.getDimensionPixelSize(R.dimen.slicePreviewDot).toFloat()
    private val lineWidth = resources.getDimensionPixelSize(R.dimen.slicePreviewWidth).toFloat()
    private val stretchedLineWidth = resources.getDimensionPixelSize(R.dimen.slicePreviewStretchedWidth).toFloat()
    private val lineDash = resources.getDimensionPixelSize(R.dimen.slicePreviewDash).toFloat()
    private val dashEffect = DashPathEffect(floatArrayOf(lineDash, lineDash), 0f)
    private var linePath: Path? = null
    private var stretchedLinePath: Path? = null
    private var startDotOval: RectF? = null
    private var endDotOval: RectF? = null


    // --
    // Initialization
    // --

    @JvmOverloads
    constructor(
        context: Context,
        attrs: AttributeSet? = null,
        defStyleAttr: Int = 0)
            : super(context, attrs, defStyleAttr)

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    constructor(
        context: Context,
        attrs: AttributeSet?,
        defStyleAttr: Int,
        defStyleRes: Int)
            : super(context, attrs, defStyleAttr, defStyleRes)

    init {
        // No implementation
    }


    // --
    // Configurable values
    // --

    var start: Point? = null
        set(start) {
            val changed = field != start
            field = start
            if (changed) {
                updateShapes()
                invalidate()
            }
        }

    var end: Point? = null
        set(end) {
            val changed = field != end
            field = end
            if (changed) {
                updateShapes()
                invalidate()
            }
        }

    var color: Int = Color.TRANSPARENT
        set(color) {
            val changed = field != color
            field = color
            if (changed) {
                invalidate()
            }
        }

    var stretchedColor: Int = Color.TRANSPARENT
        set(stretchedColor) {
            val changed = field != stretchedColor
            field = stretchedColor
            if (changed) {
                invalidate()
            }
        }


    // --
    // Update shapes
    // --

    private fun updateShapes() {
        // Create slice vectors
        val checkStart = start
        val checkEnd = end
        val sliceVector = if (checkStart != null && checkEnd != null) {
            Vector(PointF(checkStart.x.toFloat(), checkStart.y.toFloat()), PointF(checkEnd.x.toFloat(), checkEnd.y.toFloat()))
        } else {
            null
        }
        val stretchedSliceVector = sliceVector?.stretchedToEdges(PointF(0f, 0f), PointF(width.toFloat(), height.toFloat()))

        // Set up line
        linePath = if (sliceVector != null && sliceVector.isValid()) {
            Path().apply {
                moveTo(sliceVector.start.x, sliceVector.start.y)
                lineTo(sliceVector.end.x, sliceVector.end.y)
            }
        } else {
            null
        }

        // Set up stretched line
        stretchedLinePath = if (stretchedSliceVector != null && stretchedSliceVector.isValid()) {
            Path().apply {
                moveTo(stretchedSliceVector.start.x, stretchedSliceVector.start.y)
                lineTo(stretchedSliceVector.end.x, stretchedSliceVector.end.y)
            }
        } else {
            null
        }

        // Set up start/end dots
        val dotRadius = dotSize / 2
        startDotOval = null
        endDotOval = null
        start?.let {
            startDotOval = RectF(it.x - dotRadius, it.y - dotRadius, it.x + dotRadius, it.y + dotRadius)
        }
        end?.let {
            endDotOval = RectF(it.x - dotRadius, it.y - dotRadius, it.x + dotRadius, it.y + dotRadius)
        }

        // Update draw state
        setWillNotDraw(linePath == null && stretchedLinePath == null && startDotOval == null && endDotOval == null)
    }


    // --
    // Custom drawing
    // --

    override fun onDraw(canvas: Canvas?) {
        // Draw stretched line
        super.onDraw(canvas)
        stretchedLinePath?.let {
            paint.color = stretchedColor
            paint.strokeWidth = stretchedLineWidth
            paint.style = Paint.Style.STROKE
            paint.pathEffect = null
            canvas?.drawPath(it, paint)
        }

        // Draw start/end dots
        startDotOval?.let {
            paint.color = color
            paint.style = Paint.Style.FILL
            canvas?.drawOval(it, paint)
        }
        endDotOval?.let {
            paint.color = color
            paint.style = Paint.Style.FILL
            canvas?.drawOval(it, paint)
        }

        // Draw line
        linePath?.let {
            paint.color = color
            paint.strokeWidth = lineWidth
            paint.style = Paint.Style.STROKE
            paint.pathEffect = dashEffect
            canvas?.drawPath(it, paint)
        }
    }

    override fun onSizeChanged(width: Int, height: Int, oldWidth: Int, oldHeight: Int) {
        super.onSizeChanged(width, height, oldWidth, oldHeight)
        updateShapes()
    }

}
